app/scripts/modules/core/securityGroup/securityGroupReader.service.ts

```diff
@@ -0,0 +1,412 @@
+import {filter, forOwn, get, uniq} from 'lodash';
+import {module} from 'angular';
+
+import {API_SERVICE, Api} from 'core/api/api.service';
+import {NAMING_SERVICE, NamingService, IComponentName} from 'core/naming/naming.service';
+import {INFRASTRUCTURE_CACHE_SERVICE, InfrastructureCacheService} from 'core/cache/infrastructureCaches.service';
+import {Application} from 'core/application/application.model';
+import {ISecurityGroup, ILoadBalancer, ServerGroup} from 'core/domain';
+import {SECURITY_GROUP_TRANSFORMER_SERVICE, SecurityGroupTransformerService} from './securityGroupTransformer.service';
+
+interface IRegionAccount {
+  account: string;
+  accountName: string;
+  id: string;
+  name: string;
+  provider: string;
+  region: string;
+  vpcId: string;
+}
+
+interface IRegions {
+  [key: string]: IRegionAccount[]; // regionName
+}
+
+interface IProviders {
+  [key: string]: IRegions; // providerName
+}
+
+interface IGroupsByRegion {
+  [key: string]: IRegionAccount[]; // regionName
+}
+
+interface IGroupsByProvider {
+  [key: string]: IGroupsByRegion; // providerName
+}
+
+interface ISecurityGroupsByRegion {
+  [key: string]: IRegionAccount;  // regionName
+}
+
+interface ISecurityGroupsByAccount {
+  [key: string]: ISecurityGroupsByRegion;  // accountName
+}
+
+interface IIndexedSecurityGroups {
+  [key: string]: ISecurityGroupsByAccount;
+}
+
+export interface IApplicationSecurityGroup {
+  name: string;
+}
+
+interface IReaderSecurityGroup extends ISecurityGroup {
+  securityGroups: IGroupsByRegion;
+}
+
+interface ISecurityGroupDetailInboundRule {
+  accountName: string;
+  id: string;
+  inferredName: boolean;
+  name: string;
+}
+
+interface ISecurityGroupDetailRule {
+  range: any;
+  securityGroup: ISecurityGroupDetailInboundRule;
+}
+
+interface ISecurityGroupProcessorResult {
+  notFoundCaught: boolean;
+  securityGroups: ISecurityGroup[];
+}
+
+export interface IGroupsByAccount {
+  [key: string]: IProviders; // accountName
+}
+
+export interface ISecurityGroupDetail {
+  inboundRules: ISecurityGroupDetailRule[];
+  ipRangeRules: ISecurityGroupDetailRule[];
```

Comment:
These are different classes with different data. 

See:
https://github.com/spinnaker/clouddriver/blob/master/clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/securitygroups/IpRangeRule.groovy and 
https://github.com/spinnaker/clouddriver/blob/master/clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/securitygroups/SecurityGroupRule.groovy
(both implement `Rule`, defined here: https://github.com/spinnaker/clouddriver/blob/master/clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/securitygroups/Rule.groovy)